### PR TITLE
feat: add metrics per resource transformation

### PR DESCRIPTION
**Pull request by mndeveci**
This commit enables metric collection per resource, if a `Metrics` instance is been defined and configured to use CloudWatch.

*Checklist:*

- [x] Add/update tests using:
    - [x] Correct values
    - [x] Bad/wrong values (None, empty, wrong type, length, etc.)
    - [ ] [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html)
- [x] `make pr` passes
- [] Update documentation
- [ ] Verify transformed template deploys and application functions as expected

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/metrics/method_decorator.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        The function that is decorated. This will be used as metric name if name is not provided and caller is not an
+        instance of Resource object.
+    args : args
+        Arguments that is originally passed to the caller. This function will check if first element in this function
+        is a Resource then it reads the 'resource_type' property out of it to generate the metric name.
+    """
+    if name:
+        metric_name = name
+    else:
+        if args > 0 and isinstance(args[0], Resource):
```
Comment (mgrandis) [inline samtranslator/metrics/method_decorator.py:57]: Should this be `len(args) > 0`?

File: samtranslator/metrics/method_decorator.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+LOG = logging.getLogger(__name__)
+
+
+class MetricsMethodWrapperSingleton:
+    """
+    Keeps the instance of Metrics object.
+    This singleton will be alive until lambda receives shutdown event
+    """
+
+    _METRICS_INSTANCE = None
```
Comment (mgrandis) [inline samtranslator/metrics/method_decorator.py:19]: I think capital letter are reserved for constants.
This should maybe be lowercase.

File: samtranslator/model/eventsources/cloudwatchlogs.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 from samtranslator.model.types import is_str
 from samtranslator.translator.arn_generator import ArnGenerator
 from .push import PushEventSource
+from ...metrics.method_decorator import cw_timer
```
Comment (mgrandis) [inline samtranslator/model/eventsources/cloudwatchlogs.py:7]: I think absolute paths may be better

**Review by mgrandis — CHANGES_REQUESTED** [review]
A few comments

File: samtranslator/metrics/method_decorator.py (hunk 1 above)
Comment (mndeveci) [reply samtranslator/metrics/method_decorator.py:57]: Shoot, I was converting this from `if len(args) > 0` to `if args`. Let me fix this.

File: samtranslator/metrics/method_decorator.py (hunk 2 above)
Comment (mndeveci) [reply samtranslator/metrics/method_decorator.py:19]: Yes, but static instances are also using uppercase names (`LOG` for instance). I didn't want this to be accepted as instance variable that is why I went with uppercase.

**Review by CoshUS — COMMENTED** [review]
PR itself looks good to me, but while I was digging into Metrics class, I found some usage of list as default values. They should be fine with the current usages, but could lead to static variable issues in the future. Can we retrospectively change them in this PR?
https://github.com/aws/serverless-application-model/blob/develop/samtranslator/metrics/metrics.py#L122

File: samtranslator/metrics/method_decorator.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+
+def _send_cw_metric(prefix, name, execution_time_ms, func, args):
+    """
+    Gets metric name from 'prefix', 'name', 'func' and 'args' parameters, then calls metrics instance from its
+    singleton object to record the latency.
+    """
+    try:
+        metric_name = _get_metric_name(prefix, name, func, args)
+        LOG.info("Execution took %sms for %s", execution_time_ms, metric_name)
```
Comment (c2tarun) [inline samtranslator/metrics/method_decorator.py:75]: Should this be converted to `debug` to reduce some noise?

**Review by mgrandis — APPROVED** [review]

**Review by c2tarun — APPROVED** [review]

**Review by CoshUS — APPROVED** [review]
LGTM! Thanks for addressing the issue.

**Pull request merged**